packages/core/src/local-workspace/remote_map.ts

```diff
@@ -350,15 +357,81 @@ const getPrefixEndCondition = (prefix: string): string => prefix
   .substring(0, prefix.length - 1).concat((String
     .fromCharCode(prefix.charCodeAt(prefix.length - 1) + 1)))
 
+const createReadIterator = (
+  iterator: rocksdb.Iterator,
+): ReadIterator => ({
+  next: async () => readIteratorNext(iterator),
+  nextPage: async () => readIteratorPage(iterator),
+})
+
+const createFilteredReadIterator = (
+  iterator: rocksdb.Iterator,
+  filter: (x: string) => boolean,
+  limit?: number,
+): ReadIterator => {
+  let iterated = 0
+  return {
+    next: async () => {
+      const getNext = async (): Promise<remoteMap.RemoteMapEntry<string> | undefined> => {
+        if (limit !== undefined && iterated >= limit) {
+          return undefined
+        }
+        const next = await readIteratorNext(iterator)
+        if (next === undefined || filter(next.key)) {
+          iterated += 1
+          return next
+        }
+        return getNext()
+      }
+      return getNext()
+    },
+    nextPage: async () => {
+      const getNext = async (): Promise<remoteMap.RemoteMapEntry<string>[] | undefined> => {
+        if (limit && iterated >= limit) {
+          return undefined
+        }
+        const next = await readIteratorPage(iterator)
+        if (next === undefined) {
+          return undefined
+        }
+        const filteredPage = next.filter(ent => filter(ent.key))
+        if (filteredPage.length === 0) {
+          return getNext()
+        }
+        const actualPage = limit !== undefined
+          ? filteredPage.slice(0, limit - iterated)
+          : filteredPage
+        iterated += actualPage.length
+        return actualPage
+      }
+      return getNext()
+    },
+  }
+}
+
 const createIterator = (prefix: string, opts: CreateIteratorOpts, connection: rocksdb):
-rocksdb.Iterator =>
-  connection.iterator({
-    keys: opts.keys,
-    values: opts.values,
-    lte: getPrefixEndCondition(prefix),
-    ...(opts.after !== undefined ? { gt: opts.after } : { gte: prefix }),
-    ...(opts.first !== undefined ? { limit: opts.first } : {}),
-  })
+ReadIterator => (
+  opts.filter === undefined
+    ? createReadIterator(
+      connection.iterator({
+        keys: opts.keys,
+        values: opts.values,
+        lte: getPrefixEndCondition(prefix),
+        ...(opts.after !== undefined ? { gt: opts.after } : { gte: prefix }),
+        ...(opts.first !== undefined ? { limit: opts.first } : {}),
+      })
+    )
+    : createFilteredReadIterator(
+      connection.iterator({
+        keys: opts.keys,
+        values: opts.values,
+        lte: getPrefixEndCondition(prefix),
+        ...(opts.after !== undefined ? { gt: opts.after } : { gte: prefix }),
+      }),
+      opts.filter,
+      opts.first,
+    )
+)
```

Comment:
this is how I would do it.
requires checking the filter twice which isn't great, but I think it is still better than duplicating everything else...

```suggestion
ReadIterator => {
  const connectionIterator = connection.iterator({
    keys: opts.keys,
    values: opts.values,
    lte: getPrefixEndCondition(prefix),
    ...(opts.after !== undefined ? { gt: opts.after } : { gte: prefix }),
    ...(opts.first !== undefined && opts.filter === undefined ? { limit: opts.first } : {}),
  })
  return opts.filter === undefined
    ? createReadIterator(connectionIterator)
    : createFilteredReadIterator(connectionIterator, opts.filter, opts.first)
}
```

could add a more explicit part with a comment if this feels too hidden, e.g
```typescript
// Cannot use inner limit when filtering because if we filter anything out we will need
// to get additional results from the inner iterator
const limit = opts.filter !== undefined ? opts.first : undefined
const connectionIterator = connection.iterator({
  // other params
  ...(limit !== undefined ? { limit } : {}),
})
```